### PR TITLE
[MAINTENANCE] Enable explicit-override mypy rule

### PR DESCRIPTION
**Pull request by Kilo59**
Enable the `explicit-override` mypy rule.
This requires that all overridden methods must have the `@typing.override` decorator applied.

See `pep-698` for more details 👇 
https://peps.python.org/pep-0698/#limitations-of-setting-override

Due to pinned version of `typing_extensions` required by airflow `2.2` we must use `compatibility.typing_extensions` in our codebase until our min version of airflow reaches `2.5` or later.

https://gist.github.com/Kilo59/3f140fdbbcd8fb3072c289a94afef1a6

<details>
  <summary><h2>Add overrides script</h2></summary>

```python
"""
python this_script.py <PATH_TO_MODULE(s)>
"""
from __future__ import annotations

import pathlib
import sys
from collections import Counter
from pprint import pformat as pf
from typing import Iterable, Iterator, NamedTuple

from mypy import api
from typing_extensions import override


class Result(NamedTuple):
    source: pathlib.Path
    lineno: int
    message: str
    error_code: str

    @override
    def __repr__(self) -> str:
        return f"({self.source}, {self.lineno}, {self.error_code})"


def iterate_lines(stdout: str, limit: int) -> Iterator[Result]:
    for i, line in enumerate(stdout.splitlines()):
        if i >= limit:
            break
        try:
            source, lineno, _, full_message = line.split(":")
            message, error_code = full_message.split("  ")
        except ValueError:
            print(f"skipping line:\n\t{line}")
            continue
        if error_code != "[explicit-override]":
            continue
        yield Result(pathlib.Path(source), int(lineno), message.lstrip(), error_code)


LINENO_MAPPINGS: Counter[pathlib.Path] = Counter()


def insert_override(source: pathlib.Path, lineno: int) -> None:
    LINENO_MAPPINGS.update([source])
    with open(source) as f_in:
        lines = f_in.readlines()
    lines.insert(lineno - 2 + LINENO_MAPPINGS[source], "    @override\n")
    with open(source, "w") as f_out:
        f_out.writelines(lines)


def add_imports(sources: Iterable[pathlib.Path]):
    for source in sources:
        with open(source) as f_in:
            lines = f_in.readlines()
        # ruff will take care of duplicates and sorting
        lines.insert(0, "from typing_extensions import override\n")
        with open(source, "w") as f_out:
            f_out.writelines(lines)


result = api.run(sys.argv[1:])

stdout = result[0]

for error in iterate_lines(stdout, limit=1000):
    print(error)
    insert_override(error.source, error.lineno)

add_imports(LINENO_MAPPINGS.keys())

print(f"{pf(LINENO_MAPPINGS)}")
```

</details>

- [x] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [x] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [x] Code is linted - run `invoke lint` (uses `black` + `ruff`)
- [x] Appropriate tests and docs have been updated

For more information about contributing, see [Contribute](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

**Comment by ghost** [comment]
<details open>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=8589&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations">Click on the image</a> for a new way to code review</b>
</summary>

####

[Review these changes using an interactive CodeSee Map](https://app.codesee.io/r/reviews?pr=8589&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)

#### Legend
CodeSee Map legend
</details>

File: great_expectations/compatibility/typing_extensions.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+from typing import Any, Callable, TypeVar
+
+try:
+    from great_expectations.compatibility.typing_extensions import override
+except ImportError:
+    F = TypeVar("F", bound=Callable[..., Any])
+
+    def override(__arg: F, /) -> F:
+        return __arg
```
Comment (Kilo59) [inline great_expectations/compatibility/typing_extensions.py]: New `typing_extensions` compatibility module.

File: pyproject.toml
Comment (Kilo59) [inline pyproject.toml:1]: Simplified the `banned-api` config and ignored this rule for all `.pyi` stub files.

File: pyproject.toml [hunk 2]
```diff
@@ -329,6 +320,8 @@ convention = "google"
 "azure".msg = "Please do not import azure directly, import from great_expectations.compatibility.azure instead."
 "trino".msg = "Please do not import trino directly, import from great_expectations.compatibility.trino instead."
 "pyarrow".msg = "Please do not import pyarrow directly, import from great_expectations.compatibility.pyarrow instead."
+"typing_extensions.override".msg = "Do not import typing_extensions.override directly, import `override` from great_expectations.compatibility.typing_extensions instead."
+
```
Comment (Kilo59) [inline pyproject.toml:324]: Not banning all `typing_extension` imports only `.override`

**Review by cdkini — APPROVED** [review]

**Pull request merged**